WIP: prevent code generators adding testing imports

BenTheElder opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->
/kind cleanup

#### What this PR does / why we need it:

- adds code generators to verify-testing-import.sh
- cleans up misleading loop variable (file => pkg)
- ensures jq is actually available
- detects all `*testing` imports in non-test code, not just the stdlib "testing" package
- drops unnecessary exit 0
- ensures kubelet does not link testing package into prod packages via minor refactor

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes https://github.com/kubernetes/kubernetes/issues/131646

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```
/sig testing api-machinery

BenTheElder committed:
add code generators to verify-testing-import.sh

BenTheElder committed:
verify-testing-import: rename loopvar from file to pkg

we're iterating packages paths, not files

BenTheElder commented:
/cc @liggitt @jpbetz

BenTheElder commented on hack/verify-testing-import.sh:
```diff
@@ -40,21 +41,23 @@ RELEASE_BIN_PKGS=(
   "${KUBE_ROOT}/cmd/kubectl-convert"
   "${KUBE_ROOT}/cmd/kubelet"
   "${KUBE_ROOT}/cmd/kubeadm"
```
I didn't use `"${KUBE_ROOT}"/cmd/*/ in the final PR because:
- it actually doubles the time to run this (we have a lot of small utilities we don't release)
- it will fail on kubemark
- the utility seemed low

BenTheElder commented on hack/verify-testing-import.sh:
```diff
@@ -40,21 +41,23 @@ RELEASE_BIN_PKGS=(
   "${KUBE_ROOT}/cmd/kubelet"
   "${KUBE_ROOT}/cmd/kubeadm"
+  # code generators
+  "${KUBE_ROOT}"/staging/src/k8s.io/code-generator/cmd/*/
```
The trailing `/` is to avoid matching a file like OWNERS versus a directory.
For go list, it's equivalent.

I temporarily debugged the $pkg in the loop just to be sure:
```
pkg: /Users/bentheelder/go/src/k8s.io/kubernetes/cmd/cloud-controller-manager
pkg: /Users/bentheelder/go/src/k8s.io/kubernetes/cmd/kube-apiserver
pkg: /Users/bentheelder/go/src/k8s.io/kubernetes/cmd/kube-controller-manager
pkg: /Users/bentheelder/go/src/k8s.io/kubernetes/cmd/kube-proxy
pkg: /Users/bentheelder/go/src/k8s.io/kubernetes/cmd/kube-scheduler
pkg: /Users/bentheelder/go/src/k8s.io/kubernetes/cmd/kubectl
pkg: /Users/bentheelder/go/src/k8s.io/kubernetes/cmd/kubectl-convert
pkg: /Users/bentheelder/go/src/k8s.io/kubernetes/cmd/kubelet
pkg: /Users/bentheelder/go/src/k8s.io/kubernetes/cmd/kubeadm
pkg: /Users/bentheelder/go/src/k8s.io/kubernetes/staging/src/k8s.io/code-generator/cmd/applyconfiguration-gen
pkg: /Users/bentheelder/go/src/k8s.io/kubernetes/staging/src/k8s.io/code-generator/cmd/client-gen
pkg: /Users/bentheelder/go/src/k8s.io/kubernetes/staging/src/k8s.io/code-generator/cmd/conversion-gen
pkg: /Users/bentheelder/go/src/k8s.io/kubernetes/staging/src/k8s.io/code-generator/cmd/deepcopy-gen
pkg: /Users/bentheelder/go/src/k8s.io/kubernetes/staging/src/k8s.io/code-generator/cmd/defaulter-gen
pkg: /Users/bentheelder/go/src/k8s.io/kubernetes/staging/src/k8s.io/code-generator/cmd/go-to-protobuf
pkg: /Users/bentheelder/go/src/k8s.io/kubernetes/staging/src/k8s.io/code-generator/cmd/informer-gen
pkg: /Users/bentheelder/go/src/k8s.io/kubernetes/staging/src/k8s.io/code-generator/cmd/lister-gen
pkg: /Users/bentheelder/go/src/k8s.io/kubernetes/staging/src/k8s.io/code-generator/cmd/prerelease-lifecycle-gen
pkg: /Users/bentheelder/go/src/k8s.io/kubernetes/staging/src/k8s.io/code-generator/cmd/register-gen
pkg: /Users/bentheelder/go/src/k8s.io/kubernetes/staging/src/k8s.io/code-generator/cmd/validation-gen
```

liggitt commented on hack/verify-testing-import.sh:
```diff
@@ -40,21 +41,23 @@ RELEASE_BIN_PKGS=(
+for pkg in "${BIN_PKGS[@]}"
 do
-  if [ "$(go list -json "${file}" | jq 'any(.Deps[]; . == "testing")')" == "true" ]
+  if [ "$(go list -json "${pkg}" | jq 'any(.Deps[]; . == "testing")')" == "true" ]
```
in addition to not checking the code-generator packages, I think the main issue with https://github.com/kubernetes/kubernetes/issues/131646 was that we didn't catch our `k8s.io/client-go/testing` package as a "testing" package

BenTheElder replied:
I think that's only because we only checked this list of release binaries. Since the package name is "testing" it should've been caught here (like kubemark would be for cadvisor testing)

I'll verify that though.

BenTheElder replied:
I see now. Patching the jq

liggitt replied:
looks like these two:

* staging/src/k8s.io/apiserver/pkg/admission/plugin/policy/generic/policy_test_context.go, only used by tests, should move to a testing subpackage or something
* cmd/kubeadm/app/util/apiclient/dryrun.go, this will be harder to unwind, looks like it is widely used

maybe put the client-go/testing check in a separate loop with its own list of binaries to check, and exclude kubeadm from that check for now, and file an issue for cluster-lifecycle to resolve

BenTheElder replied:
kubelet also had an easy to resolve testing import https://github.com/kubernetes/kubernetes/pull/131750/commits/8d9f0f0db32fb12c05f74a8108b6f2963b785d00

BenTheElder replied:
> looks like these two:

I'm seeing even after dropping the revert commits:
```
Testing packages are imported in:
  ./cmd/cloud-controller-manager: k8s.io/client-go/testing
  ./cmd/kube-apiserver: k8s.io/client-go/testing
  ./cmd/kube-controller-manager: k8s.io/client-go/testing
  ./cmd/kube-proxy: k8s.io/client-go/testing
  ./cmd/kube-scheduler: k8s.io/client-go/testing
  ./cmd/kubectl: k8s.io/client-go/testing
  ./cmd/kubectl-convert: k8s.io/client-go/testing
  ./cmd/kubelet: k8s.io/client-go/testing
  ./cmd/kubeadm: k8s.io/client-go/testing
```

BenTheElder replied:
> staging/src/k8s.io/apiserver/pkg/admission/plugin/policy/generic/policy_test_context.go, only used by tests, easy to move to a testing subpackage or something

this is not so easy to move, there's a circular dependency problem for policy_source_test.go

liggitt replied:
hrmm... policy_source_test.go uses an isolated testing package which avoids the circular dependency

it looks like https://github.com/liggitt/kubernetes/commits/testingisolate/ gets close, but the last hangup might be use of the private guts of `type policySource` from inside `policy_test_context.go`

agree that will not be trivial to unwind, but is probably worth doing at background priority

BenTheElder replied:
Does your branch build?
You have the generictesting package using types from generic but the generic package using generictesting to test.

I tried moving the test cases but they're testing internals.

We could split into a third internal package for policy source + tests and re-export the public parts.

liggitt replied:
> Does your branch build?

no, I thought I got it down to just a couple compile issues but maybe that was masking a whole other set of issues

BenTheElder committed:
verify-testing-import.sh: detect all .*testing package imports, print which ones are being used

BenTheElder committed:
verify-testing-import.sh: require-jq

BenTheElder committed:
verify-testing-import.sh: drop unnecessary exit 0

BenTheElder commented:
Temporarily reverted of https://github.com/kubernetes/kubernetes/pull/131616, which should cause the script to fail.

BenTheElder committed:
kubelet: don't import testing dependency in non-test code

refactor NewFakeHostStatsProvider to take an injected kubecontainer.OSInterface just like NewFakeHostStatsProviderWithData

this prevents importing a testing package into non-test files, so we can enforce not linking test packages into prod binaries

BenTheElder commented:
revised failure output visible in: https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/131750/pull-kubernetes-verify/1922538970831392768

```
+++ command: bash "hack/verify-testing-import.sh"
Testing packages are imported in:
  ./cmd/cloud-controller-manager: k8s.io/client-go/testing
  ./cmd/kube-apiserver: k8s.io/client-go/testing
  ./cmd/kube-controller-manager: k8s.io/client-go/testing
  ./cmd/kube-proxy: k8s.io/client-go/testing
  ./cmd/kube-scheduler: k8s.io/client-go/testing
  ./cmd/kubectl: k8s.io/client-go/testing
  ./cmd/kubectl-convert: k8s.io/client-go/testing
  ./cmd/kubelet: k8s.io/client-go/testing, k8s.io/kubernetes/pkg/kubelet/container/testing
  ./cmd/kubeadm: k8s.io/client-go/testing
+++ exit code: 1
```

I think we should also add a more actionable message like the other linters.

dropping the temporary reverts

BenTheElder commented:
/retitle WIP: prevent code generators adding testing imports

this is going to be a little more involved, will circle back after taking care of some other things
